LogSoftMax.cu

```diff
@@ -1,173 +1,297 @@
-  {
-    __syncthreads();
-    if ((tx < stride) && (buffer[tx] < buffer[tx+stride]))
-      buffer[tx] = buffer[tx+stride];
-  }
-  if (tx == 0)
-  {
-    float max_k = -FLT_MAX;
-    if(max_k < buffer[0])
-      max_k = buffer[0];
-    buffer[LOGSOFTMAX_THREADS] = max_k;
+struct SumFloat {
+  __device__ __forceinline__ float operator()(float sum, float v) const {
+    return sum + v;
+  }
+};
+
+struct SumExpFloat {
+  __device__ __forceinline__ SumExpFloat(float v) : max_k(v) {}
+
+  __device__ __forceinline__ float operator()(float sum, float v) const {
+    return sum + expf(v - max_k);
+  }
+
+  const float max_k;
+};
+
+struct NoFinal {
+  __device__ __forceinline__ float operator()(float v) const {
+    return v;
+  }
+};
+
+struct LSMFinal {
+  __device__ __forceinline__ LSMFinal(float m) : max_k(m) {}
+
+  __device__ __forceinline__ float operator()(float v) const {
+    return max_k + logf(v);
+  }
+
+  const float max_k;
+};
+
+template <typename Reduction, typename Finalize>
+__device__ __forceinline__ float
+blockReduce(float* smem, float val,
+            const Reduction& r,
+            float defaultVal,
+            const Finalize& f) {
+  // To avoid RaW races from chaning blockReduce calls together, we
+  // need a sync here
+  __syncthreads();
+
+  smem[threadIdx.x] = val;
+
+  __syncthreads();
+
+  float warpVal = defaultVal;
+
+  // First warp will perform per-warp reductions for the remaining warps
+  if ((threadIdx.x / 32) == 0) {
+    int lane = threadIdx.x % 32;
+
+    if (lane < blockDim.x / 32) {
+#pragma unroll
+      for (int i = 0; i < 32; ++i) {
+        warpVal = r(warpVal, smem[lane * 32 + i]);
+      }
+
+      smem[lane] = warpVal;
+    }
   }
 
   __syncthreads();
 
-  // logadd?
-  float max_k = buffer[LOGSOFTMAX_THREADS];
-  buffer[tx] = 0;
-  for (int i=i_start; i<i_end; i+=i_step)
-    buffer[tx] += expf(input_k[i]-max_k);
+  // First thread will perform a reduction of the above per-warp reductions
+  float blockVal = defaultVal;
+
+  if (threadIdx.x == 0) {
+    for (int i = 0; i < blockDim.x / 32; ++i) {
+      blockVal = r(blockVal, smem[i]);
+    }
 
-  // reduce
-  for (unsigned int stride = blockDim.x >> 1; stride > 0; stride >>= 1)
-  {
-    __syncthreads();
-    if (tx < stride)
-      buffer[tx] += buffer[tx+stride];
+    smem[0] = f(blockVal);
   }
-  if (tx == 0)
-    buffer[LOGSOFTMAX_THREADS] = max_k + logf(buffer[0]);
 
+  // Sync and broadcast
   __syncthreads();
+  return smem[0];
+}
 
-  // logsoftmax
-  float logsum_k = buffer[LOGSOFTMAX_THREADS];
-  for (int i=i_start; i<i_end; i+=i_step)
-    output_k[i] = input_k[i] - logsum_k;
+template <typename Reduction>
+__device__ __forceinline__ float
+blockReduce(float* smem, float val,
+            const Reduction& r,
+            float defaultVal) {
+  return blockReduce<Reduction, NoFinal>(smem, val, r, defaultVal, NoFinal());
 }
 
+template <typename Reduction, int ILP>
```

Comment:
Look up Vasily Volkov, "Better Performance at Lower Occupancy" for a general overview of the strategy.